servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java

```diff
@@ -175,6 +205,64 @@ public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> for
+     *     each request. Instead, the balancing will happen only for every new created connection.</li>
+     *     <li>Created clients won't be able to move/shift traffic based on changes in DNS records until the remote
+     *     server closes existing connections.</li>
+     * </ol>
+     * <p>
+     * The returned builder can be customized using {@link SingleAddressHttpClientBuilderProvider}.
+     *
+     * @param host host to connect to, resolved by default using a DNS {@link ServiceDiscoverer} for every new
+     * connection. This will also be used for the {@link HttpHeaderNames#HOST} together with the {@code port}. Use
+     * {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that value
+     * or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @param port port to connect to
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
+            final String host, final int port) {
+        return forSingleAddressResolveOnDemand(HostAndPort.of(host, port));
+    }
+
+    /**
+     * Creates a {@link SingleAddressHttpClientBuilder} for an address with default {@link LoadBalancer} and DNS {@link
+     * ServiceDiscoverer} that will run resolutions before creating a new connection.
+     * <p>
+     * Important side effects to take into account:
+     * <ol>
+     *     <li>The total latency for opening a new connection will be increased by a latency of DNS resolution.</li>
+     *     <li>If the target host has more than one resolved address, created clients loose ability to load balance for
+     *     each request. Instead, the balancing will happen only for every new created connection.</li>
+     *     <li>Created clients won't be able to move/shift traffic based on changes in DNS records until the remote
+     *     server closes existing connections.</li>
+     * </ol>
+     * <p>
+     * The returned builder can be customized using {@link SingleAddressHttpClientBuilderProvider}.
+     *
+     * @param address the {@code UnresolvedAddress} to connect to, resolved using a DNS {@link ServiceDiscoverer} for
+     * every new connection. This address will also be used for the {@link HttpHeaderNames#HOST} header value.
+     * Use {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that
+     * value or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
```

Comment:
Please help with a better name 🙈

Reply:
Maybe if it is difficult to pick a name it can be worth exploring other options: we could reuse our current overloads and add one on the same name with an enum at the end i.e. (`ResolutionStrategy`) which could default to `POLL` but could be customized by the user to `ON_DEMAND`. That also leaves the door open to more customized strategies in the future if we ever need to.

Reply:
Another option, but nor sure if feasible, would be to move it into the Builder itself as a configuration option?

Reply:
I really like the `enum` idea with overloads. It makes the API clear and gives a lot of freedom to document each strategy. Thanks a lot for the suggestion!

Reply:
Actually, your idea helped to solve another problem: how users can customize SD for this discovery mode. I made the necessary adjustments in https://github.com/apple/servicetalk/pull/2520/commits/c1779b28f78b592b250b5848208f5bae29511e47 and it's possible now 🎉 
Also helped to test all edge cases.